### ISSUE TITLE
I cannot add a payment with negative value

### ISSUE THREAD

**Issue by alfonsolr09** [PrestaShop/PrestaShop#23983]
#### Describe the bug

I cannot add a payment with negative value

#### Expected behavior

in the version I had before (1.7.6.4) I could add a negative payment to rectify the invoice, now it cannot be

#### Steps to Reproduce

![image](https://user-images.githubusercontent.com/62474824/114252969-55532580-99a8-11eb-8bd9-d3d06ed182c1.png)

#### Additional information

* PrestaShop version: 1.7.7.3
* PHP version: 7.3

**Comment by sarahdib** [comment]
Hello @alfonsolr09 
Thank you for this report.

@PrestaShop/product-team what do you think about this issue ? Is it an improvement to not allow negative value ? Thank you

**Comment by alfonsolr09** [comment]
in previous versions you could enter a negative value. Like 1.7.6.4

**Comment by sarahdib** [comment]
@alfonsolr09 in the new version this page is migrated, so the previous behavior might be a bug. We need to check with product team what is the right behavior.

Thank you

**Comment by marionf** [comment]
@MatShir Do you know if it's a regression or wanted behavior ?

**Comment by alfonsolr09** [comment]
imagine that you make a return of a transfer, you should be able to put a negative value

**Comment by sarahdib** [comment]
> imagine that you make a return of a transfer, you should be able to put a negative value

 You need to do a refund using the partial refund option

**Comment by marionf** [comment]
It's not a regression and is a wanted behavior, we will study it later

**Comment by alfonsolr09** [comment]
in version 1.7.6.4 when you make a partial refund, a negative payment is not generated

**Comment by MatShir** [comment]
If the customer has already paid, it is strongly recommended doing a credit slip. If it is not already paid you can edit the order with the right amount. 
Maybe you would like an amount to be saved in the payment tab related to credit slip ? In that case, it is more a feature request

**Comment by alfonsolr09** [comment]
No, Imagine that a customer buys 2 items and makes the payment by bank transfer.
 You add a proof of payment of € 50 (total cost of the two items). But it turns out that the customer wants to return an item item. You generate a partial refund but a negative proof of payment is not added. You will continue to see that the client has paid  50€ and a negative transfer return receipt will never appear (-25€)

**Comment by alfonsolr09** [comment]
in previous versions, before migrating the order page to symfony, you could make a negative refund
![image](https://user-images.githubusercontent.com/62474824/116798732-086bf600-aaf3-11eb-82d9-f1601322eab4.png)

**Comment by matks** [comment]
> in previous versions, before migrating the order page to symfony, you could make a negative refund

We think it was _maybe_ a mistake 😄  .

We understand it seems useful, quick, very simple. But there is consequences on many aspects: invoice aspect, legal aspect, order status.

This is not something that can be decided quickly, many things must be considered to make sure
- everything works correctly
- it is legally OK 😉  I am not sure "negative payments" are law-compliant
- it is a suitable solution

I am sure the @PrestaShop/product-team will do a great work exploring this topic.

Maybe after some explorations, "adding a negative a payment" will be validated and brought back 😉  . But it needs to be verified.

**Comment by ghost** [comment]
Hello, i tried in 2020 https://github.com/PrestaShop/PrestaShop/issues/19113 :'(
I put the link to manage easier this issue

**Comment by Prestaplugins** [comment]
Until now, many payment modules were creating negative payments when you make a refund (at least Ogone / Ingenico, Payline...). On 1.7.7.x it stops working, and may cause errors on admin because these modules are still trying to do it but that's not allowed anymore.

It was good to see the negative line in admin on previous version, but right now if you allow negative value by Modifying OrderPayment class, these negative payments also appears in the original invoice giving a wrong total not corresponding to product total, because the refunded products are not displayed in the invoice, they are now displayed in the credit slip.

So a good solution would be to allow negative payment as it was already possible in the past, so you can see them in order detail, but don't display them in the invoice. Only positive payments should be displayed in the invoice.

**Comment by alfonsolr09** [comment]
Its a good idea. 
I think negative payment should be allowed as in previous versions.

**Comment by MatShir** [comment]
As @Prestaplugins said, the problem here is payment tab edits the invoice. One solution also is when creating a credit slip for a refund if the order has been paid, to automatically add the negative value to the payment tab and displays the credit slip as the document impacted without editing the invoice.

**Comment by JuanjoSCu** [comment]
It is clear that the Prestashop people do not have stores.

A customer makes an order for 3 products, payment with a card, after payment he asks you to cancel a product, you return the money on the card.
Another client asks for several things, you want to give him a discount and return the money directly on the card ...
And like those examples there are many.
You cannot tell the customer that you cancel the entire order and that another is placed.


You have to be able to add a negative payment amount, the invoice is not generated until the order leaves the store.
I understand that it cannot be modified if the invoice is generated, but before making the invoice it must be possible to add negative amounts as has been done so far.

What do I have to modify to make this work fine again?

**Comment by alfonsolr09** [comment]
+1 
is precisely what I am trying to explain. 
you are right in what you say, it shows that the developers do not have online stores

this function should work again, I would also like to know how to return it

@MatShir @matks

**Comment by matks** [comment]
Dear all,

I can read that you disagree with above decision, and there's no issue with that. It's always good to list benefits from each side of a debate.

I will just ask you to comply with our [Code of Conduct](https://github.com/PrestaShop/PrestaShop/blob/1c52cccb5ba87606a35b037ed721ff5625ffdb08/CODE_OF_CONDUCT.md):
- Being respectful of differing viewpoints and experiences
- Showing empathy towards other community members

> you are right in what you say, it shows that the developers do not have online stores

You say because we do not share your opinion we have no experience running a store. This is IMO against the `Being respectful of differing viewpoints and experiences`.

**Comment by idnovate** [comment]
You can remove these checks here:

Replace `isPrice` for `isFloat`: https://github.com/PrestaShop/PrestaShop/blob/1c52cccb5ba87606a35b037ed721ff5625ffdb08/classes/order/OrderPayment.php#L49

Comment this line: https://github.com/PrestaShop/PrestaShop/blob/1c52cccb5ba87606a35b037ed721ff5625ffdb08/src/Core/Domain/Order/Payment/Command/AddPaymentCommand.php#L105

Comment these lines: 
https://github.com/PrestaShop/PrestaShop/blob/1c52cccb5ba87606a35b037ed721ff5625ffdb08/src/PrestaShopBundle/Form/Admin/Sell/Order/OrderPaymentType.php#L113-L120

![image](https://user-images.githubusercontent.com/15104724/132327357-d1523978-e4ce-4c1b-9032-672dc527592f.png)

**Comment by alfonsolr09** [comment]
still not fixed in v 8.0....

**Comment by alfonsolr09** [comment]
Today I had to make a partial refund of a small amount on an order and it doesn't make sense that I can't add a negative payment. Without being able to add the negative payment (of the refund) I get the error that the customer has paid an amount that is no longer appropriate since I have returned the difference.
![image](https://github.com/user-attachments/assets/0d7d871c-1321-44e4-ab45-1e6d77058443)

**Comment by JuanjoSCu** [comment]
Modifica lo que dijo @idnovate, es lo que hago yo cuando hago un update y funciona perfectamente.
Es lo bueno de este código que lo puedes arreglar para que funcione bien si ........   (no añado más comentarios que se nos molestan) https://github.com/PrestaShop/PrestaShop/issues/23983#issuecomment-914176732

**Comment by alfonsolr09** [comment]
> Modifica lo que dijo [@idnovate](https://github.com/idnovate), es lo que hago yo cuando hago un update y funciona perfectamente. Es lo bueno de este código que lo puedes arreglar para que funcione bien si ........ (no añado más comentarios que se nos molestan) [#23983 (comment)](https://github.com/PrestaShop/PrestaShop/issues/23983#issuecomment-914176732)

Se podría hacer un override para que con las actualizaciones no se rompa esa modificación? O un módulo?

**Comment by idnovate** [comment]
I think that this should help:

https://devdocs.prestashop-project.org/1.7/modules/concepts/controllers/admin-controllers/override-decorate-controller/

**Comment by boo-code** [comment]
Not taking a position on whether negative payments should be allowed - that is the product decision this issue and #26542 are waiting on. One technical constraint the thread has asserted but never pinned to code, since it bears directly on that decision:

`OrderInvoice::getOrderPaymentCollection()` is `OrderPayment::getByInvoiceId($this->id)` with no filtering, and one unfiltered collection feeds two very different consumers:

- `Order::getTotalPaid()` (`classes/order/Order.php:2040`) sums it, so a negative row lowers the amount recorded as paid - the effect asked for here.
- `pdf/invoice.payment-tab.tpl:11-14` iterates the same collection and prints one line per payment with its amount, so a negative row also lands on the invoice PDF.

That is the coupling @Prestaplugins described in 2021, and it is verifiable on 9.2.x.

It means the change in #26542 - dropping `assertAmountIsPositive()` in `AddPaymentCommand`, relaxing the `isPrice` validation on `OrderPayment`, and removing the form constraint - is necessary but not sufficient. The same row that produces the wanted total also appears on the invoice document, which is where @matks' question about legal compliance actually applies: it is not the stored payment that is the problem, it is the rendered invoice.

So whichever way the decision goes, the invoice payment table needs handling of its own - either a filter there, or the credit-slip-driven variant @MatShir outlined - separately from the three validations.
